rust/src/dns/dns.rs

```diff
@@ -0,0 +1,907 @@
+   this.tx_free(tx_id);
+}
+
+#[no_mangle]
+pub extern fn rs_dns_state_tx_get(this: &mut DNSState, tx_id: libc::uint64_t)
+                                  -> *mut DNSTransaction {
+    match this.tx_get(tx_id) {
+        Some(tx) => {
+            return unsafe{transmute(tx)};
+        },
+        _ => {
+            return ptr::null_mut();
+        }
+    }
+}
+
+#[no_mangle]
+pub extern fn rs_dns_state_get_next_event(this: &mut DNSState) -> u32 {
+    match this.events.pop() {
+        Some(ev) => {
+            return ev as u32;
+        },
+        _ => {
+            return 0;
+        }
+    }
+}
+
+#[no_mangle]
+pub extern fn rs_dns_state_get_tx_count(this: &mut DNSState) -> u64 {
+    return this.tx_id;
+}
+
+#[no_mangle]
+pub extern fn rs_dns_state_parse_request(state: &mut DNSState,
+                                         input: *const libc::uint8_t,
+                                         len: libc::uint32_t) -> libc::uint64_t
+{
+    let buf = unsafe{std::slice::from_raw_parts(input, len as usize)};
+    let tx_id = state.parse_request(buf);
+    return tx_id as libc::uint64_t;
+}
+
+#[no_mangle]
+pub extern fn rs_dns_state_parse_response(state: &mut DNSState,
+                                          input: *const libc::uint8_t,
+                                          len: libc::uint32_t) -> libc::uint64_t
+{
+    let buf = unsafe{std::slice::from_raw_parts(input, len as usize)};
+    let tx_id = state.parse_response(buf);
+    return tx_id as libc::uint64_t;
+}
+
+impl DNSTransaction {
+
+    pub fn new() -> DNSTransaction {
+        return DNSTransaction{
+            id: 0,
+            request: None,
+            response: None,
+            replied: false,
+        }
+    }
+
+}
+
+#[no_mangle]
+pub extern fn rs_dns_tx_get_alstate_progress(tx: &mut DNSTransaction,
+                                             direction: libc::uint8_t)
+                                             -> libc::uint8_t {
+    if direction & core::STREAM_TOCLIENT > 0 {
+        if tx.replied {
+            return 1;
+        }
+        return 0;
+    }
+    return 1;
+}
+
+#[no_mangle]
+pub extern fn rs_dns_tx_get_query_buffer(tx: &mut DNSTransaction,
+                                         i: libc::uint16_t,
+                                         buf: *mut *const libc::uint8_t,
+                                         len: *mut libc::uint32_t)
+                                         -> libc::uint8_t
+{
+    for request in &tx.request {
+        if (i as usize) < request.queries.len() {
+            let query = &request.queries[i as usize];
+            if query.name.len() > 0 {
+                unsafe {
+                    *len = query.name.len() as libc::uint32_t;
+                    *buf = query.name.as_ptr();
+                }
+                return 1;
+            }
+        }
+    }
+    return 0;
+}
+
+#[no_mangle]
+pub extern fn rs_dns_tx_inspect_query_name(
```

Comment:
The number of answers so we know how many times to loop. I suppose it's pointless tho, it can just look at the return code which can tell it stop like it does now.